### PR TITLE
API for test suite and CLI

### LINKED ISSUE #261: Create top-level compiler API
Currently the test suite reaches into compiler internals.  We should extract that, to unblock creating a CLI.  The test suite & CLI can then use that API.

### PR DESCRIPTION
**Pull request by toothbrush**
The idea here is to pull the guts of the compiler, which are fairly intertwined with the test suite at the moment, into a "compiler driver" module to be created in the `fathom` crate.  Eventually, the hope is that this closes #261.

For now, :warning: this is still extremely messy, and will need a lot of TLC :warning:

### CODE REVIEW

File: fathom-cli/src/main.rs [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
 use structopt::StructOpt;
 
+use fathom::lang::core;
+use fathom::lang::core::binary;
+use fathom_runtime::ReadScope;
+
 mod cli;
 
-fn main() {
+fn main() -> Result<(), String> {
```
Comment (brendanzab) [inline fathom-cli/src/main.rs:10]: We could return an `anyhow::Result` here

File: fathom/src/driver/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            // test status output.
+
+            let mut buffer = BufferWriter::stderr(ColorChoice::Auto).buffer();
+            for diagnostic in &found_diagnostics {
+                term::emit(&mut buffer, &self.term_config, files, diagnostic).unwrap();
+            }
+
+            eprintln_indented(4, "", "---- found diagnostics ----");
+            eprintln_indented(4, "| ", &String::from_utf8_lossy(buffer.as_slice()));
+            eprintln!();
```
Comment (brendanzab) [inline fathom/src/driver/compiler.rs:149]: Probably don't need this kind of output for the CLI - perhaps the client of the compiler could handle the diagnostic rendering instead? So there's a `.drain_messages` method on the compiler, perhaps.

Comment (brendanzab) [reply fathom/src/driver/compiler.rs:149]: Stuff like `---- found diagnostics ----` was there for the tests to distinguish it from the regular rust/test harness output.

Comment (toothbrush) [reply fathom/src/driver/compiler.rs:149]: Fair enough!  Just so i understand, are you suggesting that the compiler simply expose the list of messages, and the CLI handle their formatting?

**Comment by brendanzab** [comment]
This was addressed as part of #283 :)

**Pull request closed without merging**